### PR TITLE
The real phylogenetics PR

### PR DESCRIPTION
**Pull request by dunleavy005**
Finished the PhyloData stuff started last year + added Amrit correction + added a new test that computes standard root-conditional phylogenetic likelihoods and compares it to an R-package solution.

### CODE REVIEW

File: SConstruct [hunk 1]
```diff
@@ -2,31 +2,51 @@ env = Environment()
...
+common_env.Append(LIBPATH=['_build/yaml-cpp', '_build/libptpll'])
+common_env.Append(LIBS=['ptpll_static', 'pll_algorithm', 'pll_optimize',
+                        'pll_tree', 'pll_util', 'pll', 'lesplace-static', 'gsl',
+                        'blas', 'yaml-cpp', 'pthread'])
 common_env.Append(LINKFLAGS=['-g'])
 
 # Doubles compilation time.
 #common_env.Append(CCFLAGS=['-O3', '-msse2'])
 
+###### FIX BELOW!!!
```
Comment (matsen) [inline SConstruct]: Note

File: src/Data.cpp [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+          .block(match_start - relpos, match_start - relpos,
+                 match_end - match_start, match_end - match_start)
+          .block(0, match_end - match_start - right_flex - 1, left_flex + 1,
+                 right_flex + 1);
+
+  return match_matrix;
+};
+
+
+/// @brief Creates the path probability matrix for the non-templated insertion
```
Comment (matsen) [inline src/Data.cpp]: Please make web link to issue explaining the logic.

File: SConstruct [hunk 3]
```diff
@@ -2,31 +2,49 @@ env = Environment()
...
+libptpll_env.VariantDir("_build/libptpll", "lib/libptpll", duplicate=0)
+libptpll_env.Command('_build/libptpll/libptpll_static.a', '',
+                     'cp -r lib/libptpll _build/ && cd _build/libptpll && ' + \
+                     'make && cp -t . _build/src/libptpll_static.a ' + \
+                     '_build/lib/_prefix/lib/libpll_algorithm.a ' + \
+                     '_build/lib/_prefix/lib/libpll_optimize.a ' + \
+                     '_build/lib/_prefix/lib/libpll_tree.a ' + \
+                     '_build/lib/_prefix/lib/libpll_util.a ' + \
+                     '_build/lib/_prefix/lib/libpll.a ' + \
+                     '_build/lib/lesplace/src/liblesplace-static.a')
```
Comment (dunleavy005) [inline SConstruct]: linearham manually builds yaml-cpp and libptpll (+ its dependencies).  Our tree structure is directly from libptpll.

File: data/PhyloData_ex/newton.fasta [hunk 4]
```diff
@@ -0,0 +1,10 @@
+>0
+TAAAAGATCAATT
+>1
+CACACGTTCGAGT
+>root
+AGGACATACGTCT
+>3
+CGTAGTAGGACTC
+
+
```
Comment (dunleavy005) [inline data/PhyloData_ex/newton.fasta]: "root" represents the naive sequence (spit out by partis, for real data examples).

File: data/PhyloData_ex/phylolikelihood_hmm_input.csv [hunk 5]
```diff
@@ -0,0 +1,2 @@
+names flexbounds relpos
+phylolikelihood "{""v_r"":[5,5],""d_r"":[10,10],""j_l"":[10,10],""j_r"":[13,13],""d_l"":[5,5],""v_l"":[0,0]}" "{""IGHD_ex*01"":5,""IGHJ_ex*01"":10,""IGHV_ex*01"":0}"
```
Comment (dunleavy005) [inline data/PhyloData_ex/phylolikelihood_hmm_input.csv]: all the `phylolikelihood`-related files are made so the full phyloHMM likelihood == standard phylo. likelihood across the MSA. we do this to test our emission calculations via libptpll.

File: src/Chain.cpp [hunk 6]
```diff
@@ -112,15 +112,60 @@ std::vector<int> Chain::ViterbiPath(int row, int col) const {
...
+void Chain::AuxFindDirtySmooshables(
+    std::vector<SmooshishPtr>& dirty_smooshables) {
+  if (is_dirty_) {
+    prev_->AuxFindDirtySmooshables(dirty_smooshables);
+    curr_->AuxFindDirtySmooshables(dirty_smooshables);
+  }
+};
+
+
 // Functions
```
Comment (dunleavy005) [inline src/Chain.cpp]: See https://github.com/matsengrp/linearham/issues/39#issuecomment-277389890 for an example of Chain cleaning.

File: src/Data.cpp (hunk 2 above)
Comment (dunleavy005) [reply src/Data.cpp]: See https://github.com/matsengrp/linearham/issues/3#issuecomment-256169007 for NTI smooshable calculation.

File: src/Data.cpp [hunk 7]
```diff
@@ -0,0 +1,609 @@
...
+  Eigen::MatrixXd match_matrix =
+      GermlineTransProbMatrix(vgerm_ptr, "v_l", "v_r");
+
+  // Multiply in the associated landing-out probabilities.
+  MultiplyLandingMatchMatrix(vgerm_ptr->landing_out(), vgerm_ptr->name(), "v_l",
+                             false, match_matrix);
+
+  // Multiply in the associated gene and padding probabilities.
+  match_matrix *= vgerm_ptr->gene_prob();
+  // double npadding_prob =
```
Comment (dunleavy005) [inline src/Data.cpp]: PhyloData npadding still needs to be addressed.

File: src/NPadding.cpp [hunk 8]
```diff
@@ -125,33 +140,37 @@ double NPadding::NPaddingProb(
...
+    flex_pad_end = std::max(read_pos, g_l);
+    n_exclude_count = std::min(flex_pad_end - read_pos, n_read_count);
   } else {
-    pad_start = std::min(read_pos, g_r);
-    pad_end = g_r;
-    n_exclude_count = std::min(read_pos - pad_start, n_read_count);
+    flex_pad_start = std::min(read_pos, g_r);
+    flex_pad_end = g_r;
+    n_exclude_count = std::min(read_pos - flex_pad_start, n_read_count);
   }
```
Comment (dunleavy005) [inline src/NPadding.cpp]: See https://github.com/matsengrp/linearham/issues/35#issuecomment-270037356 for a diagram on NPadding possibilities.

File: src/PhyloData.cpp [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+/// @brief Caches the xMSA index vector for a given germline gene.
+/// @param[in] germ_ptr
+/// A pointer to an object of class Germline.
+/// @param[in] left_flexbounds_name
+/// The name of the left flexbounds, which is a 2-tuple of MSA positions
+/// providing the bounds of the germline's left flex region.
+/// @param[in,out] xmsa_ids
+/// A map identifying already cached xMSA site indices.
+void PhyloData::CacheGermlineXmsaIndices(
+    GermlinePtr germ_ptr, std::string left_flexbounds_name,
```
Comment (dunleavy005) [inline src/PhyloData.cpp]: For some detail between MSA-xMSA conversion, see https://github.com/matsengrp/linearham/issues/39#issuecomment-272086562.

File: wercker.yml [hunk 10]
```diff
@@ -6,6 +6,17 @@ build:
             name: initialize git submodules
```
Comment (dunleavy005) [inline wercker.yml]: Random comment for Erick https://github.com/matsengrp/linearham/issues/39#issuecomment-273210340

File: test/test.cpp [hunk 11]
```diff
@@ -508,221 +471,404 @@ TEST_CASE("Smooshable", "[smooshable]") {
...
+  0.015*0.7*1*0.06,
+        0.015*0.06;
+
+  REQUIRE(ex_simple_data_ptrs[0]->vdj_pile()[0]->marginal().isApprox(V_marginal * DX_marginal * JX_marginal, 1e-5));
+  REQUIRE(ex_simple_data_ptrs[0]->vdj_pile()[1]->marginal().isApprox(V_marginal * DX_marginal * JN_nti_marginal * JN_germ_marginal, 1e-5));
+  REQUIRE(ex_simple_data_ptrs[0]->vdj_pile()[2]->marginal().isApprox(V_marginal * DN_nti_marginal * DN_germ_marginal * JX_marginal, 1e-5));
+  REQUIRE(ex_simple_data_ptrs[0]->vdj_pile()[3]->marginal().isApprox(V_marginal * DN_nti_marginal * DN_germ_marginal * JN_nti_marginal * JN_germ_marginal, 1e-5));
+  REQUIRE(ex_simple_data_ptrs[0]->MarginalLogLikelihood() == Approx(-37.3790770807));
+
+  // Test the SimpleData class using the BCRHam HMM files.
```
Comment (dunleavy005) [inline test/test.cpp]: See https://github.com/matsengrp/linearham/issues/35#issuecomment-270298028 for bcrham testing summary.

**Pull request merged**